### ISSUE TITLE
public key not found when deploying 39.20240210.3.0

### ISSUE THREAD

**Issue by akshaynshah** [coreos/fedora-coreos-tracker#1731]
### Describe the bug

Upgrading FCOS using rpm-ostree command `rpm-ostree deploy --reboot 39.20240210.3.0` is failing stating missing public key for the signature of the commit
Logs below:

### Reproduction steps

1. Create an instance from FCOS version 37.20230122.3.0
2. Try deploying FCOS version 39.20240210.3.0 using rpm-ostree deploy --reboot 39.20240210.3.0

### Expected behavior

This exact upgrade path (from 37.20230122.3.0 to 39.20240210.3.0) seemed to work couple of weeks ago but it is not working now

### Actual behavior

```
[root@localhost etc]# rpm-ostree deploy --reboot 39.20240210.3.0
Resolving version '39.20240210.3.0'
error: Commit fb67df533ff58f5770b735496e624752517f48d27e29ec0f5c47aad145afda71: Signature made Tue May  7 03:05:36 2024 using RSA key ID 0727707EA15B79CC
Can't check signature: public key not found
```
The `rpm-ostreed` journal logs:
```
May 14 16:14:27 localhost.localdomain systemd[1]: Starting rpm-ostreed.service - rpm-ostree System Management Daemon...
May 14 16:14:27 localhost.localdomain rpm-ostree[3964531]: Reading config file '/etc/rpm-ostreed.conf'
May 14 16:14:27 localhost.localdomain systemd[1]: Started rpm-ostreed.service - rpm-ostree System Management Daemon.
May 14 16:14:27 localhost.localdomain rpm-ostree[3964531]: In idle state; will auto-exit in 61 seconds
May 14 16:14:45 localhost.localdomain rpm-ostree[3964531]: Allowing active client :1.138 (uid 0)
May 14 16:14:45 localhost.localdomain rpm-ostree[3964531]: client(id:cli dbus:1.138 unit:session-10.scope uid:0) added; new total=1
May 14 16:14:45 localhost.localdomain rpm-ostree[3964531]: Loaded sysroot
May 14 16:14:45 localhost.localdomain rpm-ostree[3964531]: Locked sysroot
May 14 16:14:45 localhost.localdomain rpm-ostree[3964531]: Initiated txn Deploy for client(id:cli dbus:1.138 unit:session-10.scope uid:0): /org/projectatomic/rpmostree1/fedora_coreos
May 14 16:14:45 localhost.localdomain rpm-ostree[3964531]: Process [pid: 3964540 uid: 0 unit: session-10.scope] connected to transaction progress
May 14 16:14:46 localhost.localdomain rpm-ostree[3964531]: Txn Deploy on /org/projectatomic/rpmostree1/fedora_coreos failed: Commit fb67df533ff58f5770b735496e624752517f48d27e29ec0f5c47aad145afda71: Signature made Tue May  7 03:05:36 2024 using RSA key ID 0727707EA15B79CC
                                                           Can't check signature: public key not found
May 14 16:14:46 localhost.localdomain rpm-ostree[3964531]: Unlocked sysroot
May 14 16:14:46 localhost.localdomain rpm-ostree[3964531]: Process [pid: 3964540 uid: 0 unit: (unknown)] disconnected from transaction progress
May 14 16:14:46 localhost.localdomain rpm-ostree[3964531]: client(id:cli dbus:1.138 unit:session-10.scope uid:0) vanished; remaining=0
May 14 16:14:46 localhost.localdomain rpm-ostree[3964531]: In idle state; will auto-exit in 61 seconds
```

### System details

Platform: `AWS`
Version: 
```
[root@localhost etc]# rpm-ostree status -b
State: idle
BootedDeployment:
● fedora:fedora/x86_64/coreos/stable
                  Version: 37.20230122.3.0 (2023-02-06T21:21:08Z)
               BaseCommit: 23bfcb638fcee121d5d257349c1c964b111f7df4e9895f602e671faa9ec7a607
             GPGSignature: Valid signature by ACB5EE4E831C74BB7C168D27F55AD3FB5323552A
```

**Comment by travier** [comment]
It is not supported to skip barrier releases in Fedora CoreOS. You should let Zincati update your system through all the barrier release or use a more recent boot image.

**Comment by akshaynshah** [comment]
> It is not supported to skip barrier releases in Fedora CoreOS. You should let Zincati update your system through all the barrier release or use a more recent boot image.

Ordinarily we don't skip barrier releases but ran into some app - OS compatibility issues so couldn't update to the intermediate update in time, also this upgrade path was working until at least couple of weeks ago when we updated close to 1000 instances, not sure what happened recently.

**Comment by dustymabe** [comment]
> not sure what happened recently.

`stable` is now on Fedora Linux 40 so the latest in the OSTree repo (which is always pulled first) is signed by the Fedora 40 signing keys.

**Comment by akshaynshah** [comment]
We use Fedora CoreOS in our production environment which has approximately 4000 instances running on AWS. The problem we faced while letting zincati do the updates were

- At least from what we observed a year back the updates happened randomly so all 4000 instances running in prod could not be trusted to be on a consistent OS version. We briefly thought about using the [Fleet Lock Strategy](https://coreos.github.io/zincati/usage/updates-strategy/#lock-based-strategy) by implementing a public API controlled by us but that ran into other problems.
- It updated the system python version so all the python scripts and software running on the host OS and not in containers stopped working because the symlinks in the python virtual environment became invalid. While yes I agree that everything should run in container but in reality there is some stuff that is difficult to run in containers.
- It updated (obviously) the kernel version so the kernel driver also needed to be compiled for the updated kernel and loaded again when using [Falco](https://falco.org/docs/event-sources/kernel/) which made some of our applications that relied on falco stop working.

In addition to the above mentioned points, while 97% of the Fedora CoreOS deployment is Saas (or in cloud managed and controlled by us) some of the customers deploy our application/cloud on-premise sometimes in airgapped networks and they connect to internet like once a year in their change/upgrade window.

Hence, we decided to update the Fedora CoreOS version only when we needed it and we had updated the kernel driver and python scripts which worked with the updated OS.

--

Now to solve this problem of lagging instances (very few) and updating them to the latest version I am thinking of adding the public keys required manually to the instances, if you have any other (hack) ideas that would be of great help, cheers!

**Comment by dustymabe** [comment]
> I am thinking of adding the public keys required manually to the instances

yes, that should work to at least let the upgrade progress, but the warning from us still remains that upgrading older instances while ignoring the [update graph](https://builds.coreos.fedoraproject.org/graph?stream=stable&basearch=x86_64) is likely to cause issues for you just like this issue.
